packages/sdk-client/src/builder/ApiRootBuilder.ts

```diff
@@ -0,0 +1,131 @@
+    return this.withAuthMiddleware({
+      host: oauthUri,
+      projectKey: projectKey || this.projectKey,
+      credentials
+    })
+      .withHttpMiddleware({
+        host: baseUri,
+        fetch: fetch,
+      })
+      .withLoggerMiddleware()
+  }
+
+  withAuthMiddleware(options: AuthMiddlewareOptions): ApiRootBuilder {
+    const authMiddleware = createAuthForClientCredentialsFlow({
+      host: options.host || 'https://auth.europe-west1.gcp.commercetools.com',
+      projectKey: this.projectKey || options.projectKey,
+      credentials: {
+        clientId: options.credentials.clientId || '',
+        clientSecret: options.credentials.clientSecret || '',
+      },
+      oauthUri: options.oauthUri || '',
+      scopes: options.scopes,
+      fetch: options.fetch || fetch,
+      ...options,
+    })
+
+    this.authMiddleware = authMiddleware;
+    return this;
+  }
+
+  withHttpMiddleware(options: HttpMiddlewareOptions): ApiRootBuilder {
+    const httpMiddleware = createHttpMiddleware({
+      host: options.host || 'https://api.europe-west1.gcp.commercetools.com',
+      fetch: options.fetch || fetch,
+      ...options,
+    })
+
+    this.httpMiddleware = httpMiddleware;
+    return this;
+  }
+
+  withUserAgentMiddleware(options: UserAgentMiddlewareOptions) {
+    const userAgentMiddleware = createUserAgentMiddleware({
+      libraryName: options.libraryName || '',
+      libraryVersion: options.libraryVersion || '',
+      contactUrl: options.contactUrl || '',
+      contactEmail: options.contactEmail || '',
+      ...options
+    })
+
+    this.userAgentMiddleware = userAgentMiddleware;
+    return this;
+  }
+
+  withQueueMiddleware(options: QueueMiddlewareOptions) {
+    const queueMiddleware = createQueueMiddleware({
+      concurrency: options.concurrency,
+      ...options
+    })
+
+    this.queueMiddleware = queueMiddleware;
+    return this;
+  }
+
+  withLoggerMiddleware() {
+    const loggerMiddleware = createLoggerMiddleware();
+
+    this.loggerMiddleware = loggerMiddleware
+    return this;
+  }
+
+  withCorrelationIdMiddleware(options: CorrelationIdMiddlewareOptions) {
+    const correlationIdMiddleware = createCorrelationIdMiddleware({
+      generate: options.generate || null,
+      ...options
+    })
+
+    this.correlationIdMiddleware = correlationIdMiddleware;
+    return this;
+  }
+
+  build(): ApiRoot {
+    const { middlewares } = this;
+    if (this.authMiddleware) this.middlewares.push(this.authMiddleware);
+    if (this.httpMiddleware) this.middlewares.push(this.httpMiddleware);
+    if (this.userAgentMiddleware) this.middlewares.push(this.userAgentMiddleware);
+    if (this.correlationIdMiddleware) this.middlewares.push(this.correlationIdMiddleware);
+    if (this.loggerMiddleware) this.middlewares.push(this.loggerMiddleware);
+    if (this.queueMiddleware) this.middlewares.push(this.queueMiddleware);
```

Comment:
the order here seems to be strange and should be clarified

imho it should be more like http, queue, logger, auth, userAgent, correlationId, everything else

Reply:
I have included other authentication middleware methods `withClientCredentialsFlow`, `withPasswordFlow`, `withAnonymousSessionFlow` `withRefreshTokenFlow` etc

Now, you can use any of the 5 client authentication method.